caravan.lua

```diff
@@ -1,31 +1,311 @@
+    local type_on_count = 0
+
+    for k, type_open in ipairs(types_list) do
+        local type_on = df.global.game.main_interface.trade.current_type_a_on[list_index][k]
+        if type_on then
+            type_on_count = type_on_count + 1
+        end
+        types_list[k] = false
+    end
+
+    df.global.game.main_interface.trade.i_height[list_index] = type_on_count * 3
+end
+
+function collapseAllTypes()
+   collapseTypes(df.global.game.main_interface.trade.current_type_a_expanded[0], 0)
+   collapseTypes(df.global.game.main_interface.trade.current_type_a_expanded[1], 1)
+   -- reset scroll to top when collapsing types
+   df.global.game.main_interface.trade.scroll_position_item[0] = 0
+   df.global.game.main_interface.trade.scroll_position_item[1] = 0
+end
+
+function collapseContainers(item_list, list_index)
+    local num_items_collapsed = 0
+    for k, goodflag in ipairs(item_list) do
+        if goodflag ~= GOODFLAG.CONTAINED_UNSELECTED and goodflag ~= GOODFLAG.CONTAINED_SELECTED then
+            local next_item_index = k + 1
+            if next_item_index > #item_list - 1 then
+                goto skip
+            end
+
+            local next_item = item_list[next_item_index]
+            local this_item_is_container = df.item_binst:is_instance(df.global.game.main_interface.trade.good[list_index][k])
+            local collapsed_this_container = false
+            if this_item_is_container then
+                if goodflag == GOODFLAG.UNCONTAINED_SELECTED then
+                    item_list[k] = GOODFLAG.CONTAINER_COLLAPSED_SELECTED
+                    collapsed_this_container = true
+                elseif goodflag == GOODFLAG.UNCONTAINED_UNSELECTED then
+                    item_list[k] = GOODFLAG.CONTAINER_COLLAPSED_UNSELECTED
+                    collapsed_this_container = true
+                end
+
+                if collapsed_this_container then
+                    num_items_collapsed = num_items_collapsed + #dfhack.items.getContainedItems(df.global.game.main_interface.trade.good[list_index][k])
+                end
+            end
+
+            ::skip::
+        end
+    end
+
+    if num_items_collapsed > 0 then
+        df.global.game.main_interface.trade.i_height[list_index] = df.global.game.main_interface.trade.i_height[list_index] - num_items_collapsed * 3
+    end
+end
+
+function collapseAllContainers()
+    collapseContainers(df.global.game.main_interface.trade.goodflag[0], 0)
+    collapseContainers(df.global.game.main_interface.trade.goodflag[1], 1)
+end
+
+function collapseEverything()
+    collapseAllContainers()
+    collapseAllTypes()
+end
+
+function copyGoodflagState()
+    trader_selected_state = copyall(df.global.game.main_interface.trade.goodflag[0])
+    broker_selected_state = copyall(df.global.game.main_interface.trade.goodflag[1])
+end
+
+function CaravanOverlay:onInput(keys)
+    if not is_trade_overlay_active() then return false end
+    if keys._MOUSE_L then
```

Comment:
This will trigger multiple times in rapid succession as the mouse button is held down throughout the click. If you only want to catch the click itself, this should be `keys._MOUSE_L_DOWN`. `keys._MOUSE_L` should only be used for dragging operations.